wandb/sdk/launch/runner/kubernetes_runner.py

```diff
@@ -593,6 +607,44 @@ def inject_entrypoint_and_args(
             containers[i]["command"] = entry_point.command
 
 
+async def create_api_key_secret(
+    core_api: "CoreV1Api",
+    namespace: str,
+    api_key: str,
+) -> "V1Secret":
+    """Create a secret containing a user's wandb API key.
+
+    Arguments:
+        core_api: The Kubernetes CoreV1Api object.
+        namespace: The namespace to create the secret in.
+        api_key: The user's wandb API key
+
+    Returns:
+        The created secret
+    """
+    secret_data = {"password": base64.b64encode(api_key.encode()).decode()}
+    secret = client.V1Secret(
+        data=secret_data,
+        metadata=client.V1ObjectMeta(name="wandb-api-key", namespace=namespace),
+        kind="Secret",
+        type="kubernetes.io/basic-auth",
+    )
+
+    try:
+        try:
+            return await core_api.create_namespaced_secret(namespace, secret)
+        except ApiException as e:
+            # 409 = conflict = secret already exists
+            if e.status == 409:
+                await core_api.delete_namespaced_secret(
+                    name="wandb-api-key", namespace=namespace
+                )
+                return await core_api.create_namespaced_secret(namespace, secret)
+            raise
```

Comment:
We need to adjust our approach here. Having a universal secret for the namespace makes sense, but deleting it and creating it every time a job starts creates race conditions. Pods may fail to start because the secret doesn't exist, even though it is right about to be recreated. If we are launching multiple jobs at once this can also lead to us awaiting calls to delete and create this secret at the same time.

There is also the possibility that the secret exists but was created by an old agent and may not have the right perms for the new run.

I would recommend you just make a new secret for every run. `wandb-api-key-{run_id}` should work well.

Reply:
We could also make 1 secret per namespace but add the agent id to it, if applicable, if there are concerns about the # of secrets we would be creating/deleting.

Reply:
Cool, thanks for the suggestion. I changed it to use the agent ID, and manually tested to make sure it still works.

Reply:
I think we need to add back some cleanup for the secrets, we shouldn't orphan any k8s resources. I think the cleanup part for the agent secrets may be tricky, because you would have to keep track of the namespaces you have added a secret to and then only delete that secret once the agent is exiting or maybe once there are no jobs running in that namespace. The cleanup is probably easier if you do it one per run id because then can cleanup the secret at the end.

Reply:
And I also see now that we don't clean up the pull secret

Reply:
Actually I guess the cleanup is tricky in either case because `run` doesn't necessarily wait for the job to starts and that could be up to 30 minutes after `run` finishes. Or what if a user launches on the cli and sets `--async` - then it seems like it could be impossible for us to clean up any secrets that would created.

Reply:
Handling cleanup is tricky in the non-helm case, but to handle cleanup for agents deployed using helm, I've made a change to the [helm chart](https://github.com/wandb/helm-charts/pull/59) so that all of the api key secrets used by an agent are the helm-created ones, when possible.